### PR TITLE
Reduce memory footprint for Val type

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

**Comment by dsyme** [comment]
@AviAvni I can't for the life of me see why #d24532d made such a difference. Odd.  Perhaps it's a CI thing, but all the jobs fail consistently...

**Comment by dsyme** [comment]
@OK, I think I found it.  The problem is in SetData. This part:

```fsharp
        match x.val_opt_data,tg.val_opt_data with
        | Some x, Some tg ->
            x.val_compiled_name   <- tg.val_compiled_name
            x.val_other_range     <- tg.val_other_range  
            x.val_const           <- tg.val_const        
            x.val_defn            <- tg.val_defn         
            x.val_member_info     <- tg.val_member_info  
            x.val_repr_info       <- tg.val_repr_info    
        | Some _, None -> x.val_opt_data <- None
        | None, Some tg -> x.val_opt_data <- Some({ val_compiled_name = tg.val_compiled_name; val_other_range = tg.val_other_range; val_const = tg.val_const; val_defn = tg.val_defn; val_member_info = tg.val_member_info; val_repr_info = tg.val_repr_info })
        | None, None -> ()
```
should be just
```fsharp
        match tg.val_opt_data with
        | Some tg ->
            x.val_opt_data <- Some({ val_compiled_name = tg.val_compiled_name; val_other_range = tg.val_other_range; val_const = tg.val_const; val_defn = tg.val_defn; val_member_info = tg.val_member_info; val_repr_info = tg.val_repr_info })
        | None -> ()
```

If there was already a val_opt_data node in the data then we should nuke it, we don't want to recycle the old storage.  

I'm not totally sure where this makes a difference - surely the old val_data would need to be accessible for this to matter - but I presume this makes a difference here:
```
and fixupValData g compgen tmenv (v2:Val) =
    // only fixup if we copy the value
    match compgen with 
    | OnlyCloneExprVals when v2.IsMemberOrModuleBinding -> ()
    | _ ->  
        let newData = remapValData g tmenv v2 |> markAsCompGen compgen
        // uses the same stamp
        v2.SetData newData
    
```

**Comment by AviAvni** [comment]
@dsyme Thanks now it works I'm continuing to work on the PR

File: src/fsharp/TastOps.fs [hunk 1]
```diff
@@ -4632,14 +4632,23 @@ and remapValReprInfo g tmenv (ValReprInfo(tpNames, arginfosl, retInfo)) =
...
-        val_attribs       = d.val_attribs       |> remapAttribs g tmenv }
+        val_type     = ty';
+        val_opt_data =
+            match d.val_opt_data with
+            | Some dd ->
+                Some { dd with 
+                         val_repr_info = val_repr_info
+                         val_member_info = val_member_info
+                         val_declaring_entity = val_declaring_entity }
+            | None -> None;
```
Comment (forki) [inline src/fsharp/TastOps.fs]: please don't introduce new ;

Comment (forki) [reply src/fsharp/TastOps.fs]: Also Option.map should be valid here

File: src/fsharp/TastPickle.fs [hunk 2]
```diff
@@ -1812,20 +1812,20 @@ and p_vrefFlags x st =
 
 and p_ValData x st =
     p_string x.val_logical_name st
-    p_option p_string x.val_compiled_name st
+    p_option p_string (match x.val_opt_data with | Some x -> x.val_compiled_name | _ -> None) st
     // only keep range information on published values, not on optimization data
-    p_ranges (if x.val_repr_info.IsSome then Some(x.val_range, x.DefinitionRange) else None) st
+    p_ranges (if x.ValReprInfo.IsSome then Some(x.val_range, x.DefinitionRange) else None) st
```
Comment (forki) [inline src/fsharp/TastPickle.fs]: can we get rif of isSome? Option.map?

File: src/fsharp/tast.fs [hunk 3]
```diff
@@ -2216,29 +2180,68 @@ and [<StructuredFormatDisplay("{LogicalName}")>]
...
       // MUTABILITY CLEANUP: mutability of this field is used by 
       //     -- LinearizeTopMatch
       //
       // The fresh temporary should just be created with the right parent
       mutable val_declaring_entity: ParentRef
+    }
+and ValData = Val
+and [<StructuredFormatDisplay("{LogicalName}")>]
+    Val = 
+    // ValData is 19 words!! CONSIDER THIS TINY FORMAT, for all local, immutable, attribute-free values
```
Comment (forki) [inline src/fsharp/tast.fs]: is the comment still true?

**Comment by AviAvni** [comment]
@dsyme I added the measurement we talked about https://github.com/AviAvni/visualfsharp/commit/33fa71e164b53fa391d427de8f640e97a6f6efda
how can I get it when building the compiler?
I can see the stats on some tests like in access-FSC_BASIC
```
~~~ Val statistics: WithOptData 3863, WithoutOptData 18328
```
nbody
```
~~~ Val statistics: WithOptData 2214, WithoutOptData 7070
```

**Comment by AviAvni** [comment]
@dsyme I found a way to do it on the compiler
```
~~~ Val statistics: WithOptData 122009, WithoutOptData 1680496
```

File: src/fsharp/TastPickle.fs [hunk 4]
```diff
@@ -2121,23 +2121,31 @@ and u_ValData st =
...
+    let res =
+        { val_logical_name=x1
+          val_range=(match x1a with None -> range0 | Some(a,_) -> a)
+          val_type=x2
+          val_stamp=newStamp()
+          val_flags=ValFlags(x4)
+          val_opt_data=None
+        }
+
+    res.val_opt_data <-
```
Comment (forki) [inline src/fsharp/TastPickle.fs]: Instead of the mutation you could set it directly here, right?

File: src/fsharp/tast.fs [hunk 5]
```diff
@@ -2216,29 +2180,71 @@ and [<StructuredFormatDisplay("{LogicalName}")>]
...
+      mutable val_opt_data : ValOptionalData option } 
+
+    static member EmptyValOptData = { val_compiled_name = None; val_other_range = None; val_const = None; val_defn = None; val_repr_info = None; val_access = TAccess []; val_xmldoc = XmlDoc [||]; val_member_info = None; val_declaring_entity = ParentNone; val_xmldocsig = String.Empty; val_attribs = [] }
 
     /// Range of the definition (implementation) of the value, used by Visual Studio 
     member x.DefinitionRange            = 
-        match x.val_other_range with
-        | Some (m,true) -> m
+        match x.val_opt_data with
+        | Some ({ val_other_range = Some(m,true) }) -> m
```
Comment (forki) [inline src/fsharp/tast.fs]: () can go away

File: src/fsharp/tast.fs [hunk 6]
```diff
@@ -2216,29 +2180,71 @@ and [<StructuredFormatDisplay("{LogicalName}")>]
...
         | _ -> x.val_range
 
     /// Range of the definition (signature) of the value, used by Visual Studio 
     member x.SigRange            = 
-        match x.val_other_range with
-        | Some (m,false) -> m
+        match x.val_opt_data with
+        | Some ({ val_other_range = Some(m,false) }) -> m
```
Comment (forki) [inline src/fsharp/tast.fs]: () can go away

File: src/fsharp/tast.fs [hunk 7]
```diff
@@ -2311,13 +2326,19 @@ and [<StructuredFormatDisplay("{LogicalName}")>]
     member x.IsExtensionMember          = x.val_flags.IsExtensionMember
 
     /// The quotation expression associated with a value given the [<ReflectedDefinition>] tag
-    member x.ReflectedDefinition        = x.val_defn
+    member x.ReflectedDefinition        =
+        match x.val_opt_data with
+        | Some (x) -> x.val_defn
```
Comment (forki) [inline src/fsharp/tast.fs]: () can go away

File: src/fsharp/tast.fs [hunk 8]
```diff
@@ -4981,23 +5020,31 @@ let NewModuleOrNamespace cpath access (id:Ident) xml attribs mtype = Construct.N
...
+    let res =
+        Val.New
+            { val_stamp = stamp
+              val_logical_name=logicalName
+              val_range=m
+              val_flags = ValFlags(recValInfo,baseOrThis,isCompGen,inlineInfo,isMutable,isModuleOrMemberBinding,isExtensionMember,isIncrClassSpecialMember,isTyFunc,allowTypeInst,isGeneratedEventVal)
+              val_type = ty
+              val_opt_data = None } 
+
+    res.val_opt_data <-
```
Comment (forki) [inline src/fsharp/tast.fs]: same here. I think you can set it directly without mutation here

File: src/fsharp/TastPickle.fs (hunk 4 above)
Comment (AviAvni) [reply src/fsharp/TastPickle.fs]: Yes I can but maybe @dsyme  wanted to do this in a way that it's done in tast.fs

Comment (forki) [reply src/fsharp/TastPickle.fs]: was just an idea. I think it would cleanup the code here nicely

File: src/fsharp/TastOps.fs [hunk 9]
```diff
@@ -4632,14 +4632,23 @@ and remapValReprInfo g tmenv (ValReprInfo(tpNames, arginfosl, retInfo)) =
 
 and remapValData g tmenv (d: ValData) =
     let ty = d.val_type
-    let topValInfo = d.val_repr_info
+    let topValInfo = d.ValReprInfo
     let ty' = ty |> remapPossibleForallTy g tmenv
+    let val_declaring_entity = d.DeclaringEntity |> remapParentRef tmenv
```
Comment (dsyme) [inline src/fsharp/TastOps.fs]: change these names, e.g. `let declaringEntity' = ` or some other systematic naming convention to indicate old/new values in the remap.  I don't particularly like `'` so you can use `let declaringEntityR` or something where `R` means `remapped` etc

File: src/fsharp/tast.fs [hunk 10]
```diff
@@ -2216,29 +2180,71 @@ and [<StructuredFormatDisplay("{LogicalName}")>]
...
+
+      /// Custom attributes attached to the value. These contain references to other values (i.e. constructors in types). Mutable to fixup  
+      /// these value references after copying a collection of values. 
+      mutable val_attribs: Attribs
+    }
+
+and ValData = Val
+and [<StructuredFormatDisplay("{LogicalName}")>]
+    Val = 
+    // ValData is 19 words!! CONSIDER THIS TINY FORMAT, for all local, immutable, attribute-free values
```
Comment (dsyme) [inline src/fsharp/tast.fs]: Remove this comment

File: src/fsharp/tast.fs [hunk 11]
```diff
@@ -2311,13 +2326,19 @@ and [<StructuredFormatDisplay("{LogicalName}")>]
     member x.IsExtensionMember          = x.val_flags.IsExtensionMember
 
     /// The quotation expression associated with a value given the [<ReflectedDefinition>] tag
-    member x.ReflectedDefinition        = x.val_defn
+    member x.ReflectedDefinition        =
+        match x.val_opt_data with
+        | Some x -> x.val_defn
```
Comment (dsyme) [inline src/fsharp/tast.fs]: Don't re-use the name `x` in all of these - just call it `optData` or something :)

File: src/fsharp/tast.fs [hunk 12]
```diff
@@ -2401,18 +2422,27 @@ and [<StructuredFormatDisplay("{LogicalName}")>]
...
+        match x.val_opt_data with
+        | Some x -> x.val_xmldoc
+        | _ -> XmlDoc.Empty
     
     ///Get the signature for the value's XML documentation
     member x.XmlDocSig 
-        with get() = x.val_xmldocsig
-        and set(v) = x.val_xmldocsig <- v
+        with get() = match x.val_opt_data with | Some x -> x.val_xmldocsig | _ -> String.Empty
+        and set(v) = match x.val_opt_data with | Some x -> x.val_xmldocsig <- v | _ -> x.val_opt_data <- Some { Val.EmptyValOptData with val_xmldocsig = v }
```
Comment (dsyme) [inline src/fsharp/tast.fs]: Make this a multi-line expression.  There are just too many `<-`, `->` etc. on the same line :)

File: src/fsharp/tast.fs [hunk 13]
```diff
@@ -4980,24 +5019,29 @@ exception FullAbstraction of string * range
...
+        { val_stamp        = stamp
+          val_logical_name = logicalName
+          val_range        = m
+          val_flags        = ValFlags(recValInfo,baseOrThis,isCompGen,inlineInfo,isMutable,isModuleOrMemberBinding,isExtensionMember,isIncrClassSpecialMember,isTyFunc,allowTypeInst,isGeneratedEventVal)
+          val_type         = ty
+          val_opt_data     =
+            match compiledName, arity, konst, access, doc, specialRepr, actualParent, attribs with
+            | None, None, None, TAccess [], XmlDoc [||], None, ParentNone, [] -> None
+            | _ -> 
+                Some { val_compiled_name    = (match compiledName with Some v when v <> logicalName -> compiledName | _ -> None)
```
Comment (dsyme) [inline src/fsharp/tast.fs]: You could use `{ ValOptData.Empty with ... }` here to avoid the initializing of some of the fields

**Comment by dsyme** [comment]
@dotnet-bot Test Ubuntu16.04 Release_default Build please

**Comment by dsyme** [comment]
@AviAvni This looks great, I made a few minor comments

I do think we need to validate somehow that the "thin" values are being used in the majority of cases.  I would recommend doing that by manually augmenting the code to collect statistics, and then report those statistics on application exit by printing to console or writing to a log file.  The code can be temporary one-off.  Then compile-the-compiler and see the stats

File: src/fsharp/TastOps.fs (hunk 9 above)
Comment (forki) [reply src/fsharp/TastOps.fs]: yes was going to suggest the same the tick makes it often disappear in debugger

**Comment by dsyme** [comment]
OK, fantastic! I didn't see the stats - that's excellent news, 10% are big, the rest are small.  Perfect!

I'll label as approved.

**Comment by dsyme** [comment]
Looking at [these old figures](http://fsharp.github.io/2015/09/29/fsharp-compiler-guide.html#compiler-memory-usage), ValData accounts for about 12% of long-term memory usage.  This change appears to reduce the size of ValData to about 40% of what it was before for about 93% of ValData values, and a 5% increase for about 7%. 

If those figures are still accurate the change would give a total long-lived memory usage reduction in the managed-memory heap of 12 * (0.93 * 0.40 + 0.07 * 1.05) = 5.4%.

**Comment by forki** [comment]
Simply amazing.

Am 28.02.2018 17:17 schrieb "Don Syme" <notifications@github.com>:

**Comment by AviAvni** [comment]
@dsyme Done

**Comment by KevinRansom** [comment]
@AviAvni 
Thanks for this

Kevin

**Pull request merged**